frontend/apps/reader/modules/readertoc.lua

```diff
@@ -215,6 +215,34 @@ function ReaderToc:getTocIndexByPage(pn_or_xp)
     return pre_index
 end
 
+function ReaderToc:getAccurateTocIndexByXPointer(xptr)
+    local pageno = self.ui.document:getPageFromXPointer(xptr)
+    local index = self:getTocIndexByPage(pageno) -- get toc entry on the page
```

Comment:
It's a bit confusing here. Looking at that, it seems you could get in an infinite loop with:
getTocIndexByPage calling getAccurateTocIndexByXPointer callling getTocIndexByPage calling getAccurateTocIndexByXPointer ...
I guess you get out of that infinite loop because on first call you provide the xp string, and on the 2nd call, you provide a pageno integer.
But it's quite confusing :)
So, may be there's a clearer way to do that, and if not, please add some comments about why doing it that way without risk.

Except for that, I'm happy with the refactoring :)